### PR TITLE
Expand ForwardDiff support to LocalNonlinearity parameters

### PR DESCRIPTION
**Pull request by niklasschmitz**
This PR modifies the generic LocalNonlinearity term to allow for ForwardDiff.Dual numbers inside the term. This enables differentiation of e.g. forces w.r.t. parameters inside LocalNonlinearity as shown in the new testcase. Possible follow-up work would be going along these lines for other terms too.

### CODE REVIEW

File: src/terms/local_nonlinearity.jl [hunk 1]
```diff
@@ -22,14 +23,14 @@ function ene_ops(term::TermLocalNonlinearity, basis::PlaneWaveBasis{T}, ψ, occu
 end
 
 
-function compute_kernel(term::TermLocalNonlinearity, ::AbstractBasis; ρ, kwargs...)
```
Comment (mfherbst) [inline src/terms/local_nonlinearity.jl:25]: `::AbstractBasis{T}`

File: test/forwarddiff.jl [hunk 2]
```diff
@@ -130,3 +130,31 @@ end
...
+@testset "LocalNonlinearity sensitivity using ForwardDiff" begin
+    function compute_force(ε::T) where {T}
+        # solve the 1D Gross-Pitaevskii equation with ElementGaussian potential
+        a = 10.0
+        lattice = a .* [[1 0 0.]; [0 0 0]; [0 0 0]]
+        positions = [[0.2, 0, 0], [0.8, 0, 0]]
+        gauss = ElementGaussian(1.0, 0.5)
+        atoms = [gauss, gauss]
+        C = 1.0
+        α = 2
```
Comment (mfherbst) [inline test/forwarddiff.jl:143]: It's a test ... just hard code these values.

File: test/forwarddiff.jl [hunk 3]
```diff
@@ -130,3 +130,31 @@ end
...
+        n_electrons = 1
+        terms = [Kinetic(), AtomicLocal(), LocalNonlinearity(ρ -> (C + ε) * ρ^α)]
+        model = Model(Matrix{T}(lattice), atoms, positions;
+                      n_electrons, terms, spin_polarization=:spinless)
+        basis = PlaneWaveBasis(model; Ecut=500, kgrid=(1, 1, 1))
```
Comment (mfherbst) [inline test/forwarddiff.jl:148]: How long does this take ... might be ok, but perhaps you can go lower here.

File: src/terms/local_nonlinearity.jl (hunk 1 above)
Comment (niklasschmitz) [reply src/terms/local_nonlinearity.jl:25]: [080a1f7](https://github.com/JuliaMolSim/DFTK.jl/pull/779/commits/080a1f776820c71b32062ef47e7d00268f13547b)
and
[5f41cad](https://github.com/JuliaMolSim/DFTK.jl/pull/779/commits/5f41cad0d42f469334cb0dbbeacf90975e21b358)

File: test/forwarddiff.jl (hunk 2 above)
Comment (niklasschmitz) [reply test/forwarddiff.jl]: [6a679f5](https://github.com/JuliaMolSim/DFTK.jl/pull/779/commits/6a679f54db4441f1e6c26e88546555fca7368e7d)

File: src/terms/local_nonlinearity.jl [hunk 4]
```diff
@@ -13,6 +13,7 @@ function ene_ops(term::TermLocalNonlinearity, basis::PlaneWaveBasis{T}, ψ, occu
                  ρ, kwargs...) where {T}
     fp(ρ) = ForwardDiff.derivative(term.f, ρ)
     E = sum(term.f.(ρ)) * basis.dvol
+    E = convert_dual(T, E)
```
Comment (mfherbst) [inline src/terms/local_nonlinearity.jl:16]:
```suggestion
    E = sum(fρ -> convert_dual(T, fρ), term.f.(ρ)) * basis.dvol
```
has the advantage of doing less unneccessary computation in dual numbers and writing everything as one line.

**Comment by mfherbst** [comment]
Tests fail because of https://github.com/JuliaMolSim/PseudoLibrary/issues/7

**Pull request merged**